feat(docs) Add New Behavior Guide

kurtis-lew opened the pull request:
Most recently discussed on Discord at:
https://discord.com/channels/719497620560543766/719544230497878116/952829891999576134

This PR adds a guide for developers wishing to create new behaviors for ZMK. The behavior guide is largely inspired by the new shield guide, and shows users how to:

1. Create the behavior
2. Test changes locally
3. Document behavior functionality
4. Create a Pull Request for review and inclusion into the ZMK sources

kurtis-lew commented:
Made edits in 1cbea31 based on conversations with @dnaq on Discord.
The following suggestions provide a basis for these updates and were posted with their permission to allow for further discussion and/or elaboration on what the implementation of these ideas may look like.

> 1. More discussion and examples of the zmk api calls, maybe also some discussion about the event system
> 2. A walkthrough of implementing a simple, but non-trivial behaviour
>     Something that takes a bit of configuration, maybe uses a timer. Maybe something like a window switcher.

henrebotha commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,358 @@
+
+Allows ZMK to assign the correct driver to the behavior extracted from the keymap or `.dtsi`.
+The `compatible` property should have the same value as the name of the `.yaml` file
+surrounded by quotation marks. As shown above, `compatible: "zmk,behavior-hold-tap" is the value of the `compatible`property of`zmk,behavior-hold-tap.yaml`.
```
```suggestion
surrounded by quotation marks. As shown above, `compatible: "zmk,behavior-hold-tap"` is the value of the `compatible` property of `zmk,behavior-hold-tap.yaml`.
```

kurtis-lew commented:
Fixed several of the links to the relevant Zephyr documentation. Currently, the main areas these docs are lacking in are:

- ZMK Event Manager: Missing description of the `return` values.
- API Structure -> `.locality`: Missing description of the `BEHAVIOR_LOCALITY_<value>`s.
- Data Pointers: Data gets assigned the value of `dev->data` for some `const struct device *dev`. Unsure how to describe what the `dev` actually is.

kurtis-lew commented:
> * Data Pointers: Data gets assigned the value of `dev->data` for some `const struct device *dev`. Unsure how to describe what the `dev` actually is.

Updated section in b0756b3 according to my understanding.

kurtis-lew commented:
Added locality descriptions to the API Structure section in cd2fafb. With this update, the PR feels complete enough to open for further review.

kurtis-lew commented:
Posting public preview!
https://deploy-preview-1174--zmk.netlify.app/docs/development/new-behavior

dxmh reviewed (commented):
This is an excellent resource @kurtis-lew! 

I've taken a look from a docs/reader point of view and left a few minor comments below, but I think it's great overall and it'll be a very useful addition to the docs. 👍 

I think it would be good to get approval from at least one developer contributor familiar with the more technical aspects. Maybe @petejohanson, @joelspadin or @okke-formsma?

kurtis-lew commented:
Thanks for the review, @dxmh! I've committed and squashed all the little things you've suggested so far. Any additions you requested (are `description`s in the DT bindings for end-users or devs, removing the link from `DEVICE_DT_INST_DEFINE` and reinserting it into the main content) have also been pushed. I also went ahead and fixed some of the links to the other behavior pages (mod-tap/hold-tap).

I definitely agree with having any of the developer contributors share their input. Looking forward to making more progress on this PR!

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+    - `zmk_behavior_binding`: Stores the name of the behavior device (`char *behavior_dev`) as a `string` and up to two additional parameters (`uint32_t param1`, `uint32_t param2`)
+    - `zmk_behavior_binding_event`: Contains layer, position, and timestamp data for an active `zmk_behavior_binding`
+
+Other common dependencies include `zmk/keymap.h`, which allows behaviors to access layer information and extract behavior bindings from keymaps, and `zmk/event_manager.h`.
```
```suggestion
Other common dependencies include `zmk/keymap.h`, which allows behaviors to access layer information and extract behavior bindings from keymaps, and `zmk/event_manager.h` which is detailed below.
```

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+
+Comes in the form `static const struct behavior_driver_api <behavior_name>_driver_api)`. Common items to include in the API Structure are:
+
+- `.binding_pressed`: Used for behaviors that invoke an action on its keybind press. Set `.binding_pressed` equal to the function typically named [`on_<behavior_name>_binding_pressed.`](#dependencies)
```
```suggestion
- `.binding_pressed`: Used for behaviors that invoke an action on its keybind press. Set `.binding_pressed` equal to the function typically named [`on_<behavior_name>_binding_pressed`](#dependencies).
```

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+Comes in the form `static const struct behavior_driver_api <behavior_name>_driver_api)`. Common items to include in the API Structure are:
+
+- `.binding_pressed`: Used for behaviors that invoke an action on its keybind press. Set `.binding_pressed` equal to the function typically named [`on_<behavior_name>_binding_pressed.`](#dependencies)
+- `.binding_released`: Same as above, except for activating on keybind release events. Set `.binding_released` equal to the function typically named [`on_<behavior_name>_binding_released.`](#dependencies)
```
```suggestion
- `.binding_released`: Same as above, except for activating on keybind release events. Set `.binding_released` equal to the function typically named [`on_<behavior_name>_binding_released`](#dependencies).
```

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+- `.binding_pressed`: Used for behaviors that invoke an action on its keybind press. Set `.binding_pressed` equal to the function typically named [`on_<behavior_name>_binding_pressed.`](#dependencies)
+- `.binding_released`: Same as above, except for activating on keybind release events. Set `.binding_released` equal to the function typically named [`on_<behavior_name>_binding_released.`](#dependencies)
+- `.locality`: Defined in `<drivers/behavior.h>`. Describes how the behavior affects parts of a _split_ keyboard.
+  - `BEHAVIOR_LOCALITY_CENTRAL`: Behavior only affects the central half
+  - `BEHAVIOR_LOCALITY_EVENT_SOURCE`: Behavior affects only the central _or_ peripheral half, depending on which side invoked the behavior binding
+  - `BEHAVIOR_LOCALITY_GLOBAL`: Behavior affects the entire keyboard
```
```suggestion
  - `BEHAVIOR_LOCALITY_CENTRAL`: Behavior only affects the central half, which is the case for most keymap-related behavior.
  - `BEHAVIOR_LOCALITY_EVENT_SOURCE`: Behavior affects only the central _or_ peripheral half depending on which side invoked the behavior binding, such as [reset behaviors](../behaviors/reset.md).
  - `BEHAVIOR_LOCALITY_GLOBAL`: Behavior affects the entire keyboard, such as [external power](../behaviors/power.md) and lighting-related behaviors that need to be synchronized across halves.
```
Can we add some comments like above to describe what types of behaviors need what kind of locality?

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+
+### Updating `app/CmakeLists.txt` to include the new driver
+
+Behavior drivers are stored after the line `if ((NOT CONFIG_ZMK_SPLIT) OR CONFIG_ZMK_SPLIT_BLE_ROLE_CENTRAL)` in the form, `app PRIVATE src/behaviors/<behavior_name>.c`, as shown below.
```
I think this depends on the locality right? This seems to apply to central locality behaviors.

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+
+The format of a behavior's `.dtsi` file is identical to declaring an instance of the behavior in a user's keymap. The only major difference is that the value `/omit-if-no-ref/` should be placed adjacent to the name of the behavior, as seen in line 11 of the `gresc` example.
+
+## Testing changes Locally
```
```suggestion
## Testing changes locally
```

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+See [Documentation](documentation.md) for more information on writing, testing, and formatting ZMK documentation.
+:::
+
+## Submitting a Pull Request
```
```suggestion
## Submitting a pull request
```

caksoylar commented on docs/docs/development/new-behavior.md:
````diff
@@ -0,0 +1,383 @@
+Using `zmk,behavior-hold-tap.yaml` as a reference:
+
+```
+# Copyright (c) 2020 The ZMK Contributors
````
Should we keep the copyright headers here and in the `gresc` example? Unless it is needed for license purposes removing them would make this page more concise.

kurtis-lew replied:
Developers that are new to ZMK might not know to sign the files they create with the copyright headers for consistency across the codebase. Since Devicetree bindings, drivers, and pre-defined instances need to be signed with our `Copyright (c) XXXX The ZMK Contributors` headers, it would probably be useful to include some mention of the copyright stuff somewhere.

caksoylar replied:
That's true, it's probably worth pointing them out somewhere explicitly. Maybe we can add it to the last section (opening a pull request)? 

For these examples, we could say "[...] shown below as a reference (omitting copyright headers):"? But in any case, it's a minor issue and I don't mind the headers staying there either.

kurtis-lew replied:
I think the pull request section is a great place to put the copyright information. It would also help to keep things more concise, perhaps by implementing a tabbed menu that lets users copy and paste the correct copyright header depending on each file. Something similar to this section of the new shield guide:
![image](https://user-images.githubusercontent.com/27038994/165007225-cb68a59a-d823-4b94-87f4-44a20de6d7c1.png)

kurtis-lew replied:
Resolved in [aa17e24](https://github.com/zmkfirmware/zmk/commit/aa17e247bb23a6fefaf00be34aedf3ea376077ba)

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+The fifth cell of `DEVICE_DT_INST_DEFINE` can be set to `NULL` instead if instance-specific configurations are not required.
+
+:::caution
+Remember that `.c` files should be formatted according to `clang-format` to ensure that checks run smoothly once the Pull Request is submitted.
```
```suggestion
Remember that `.c` files should be formatted according to `clang-format` to ensure that checks run smoothly once the pull request is submitted.
```

caksoylar commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,383 @@
+- `zmk/events/keycode_state_changed.h`: Keycode events' state (on/off), usage page, keycode value, modifiers, and timestamps
+- `zmk/events/modifiers_state_changed.h`: Modifier events' state (on/off) and modifier value
+
+Events can be used similarly to hardware interrupts.
```
I think it'd be great to include something here for subscriptions and listeners. Listeners are referred to below but they are not introduced. Maybe we could at least mention and add a link to Zephyr docs?

kurtis-lew replied:
Resolved in [aa17e24](https://github.com/zmkfirmware/zmk/commit/aa17e247bb23a6fefaf00be34aedf3ea376077ba)

caksoylar reviewed (commented):
Some questions and a few style nits. Thank you for working on this, it is going to be a great resource!

henrebotha commented on docs/docs/development/new-behavior.md:
```diff
@@ -0,0 +1,381 @@
+`return` values:
+
+- `ZMK_EV_EVENT_BUBBLE`: The function saw the event memory and is done using it. Keep propagating the event memory to the next listeners.
+- `ZMK_EV_EVENT_HANDLED`: The function received the event memory and has handled it, stop propagating it to the next listeners and free the event payload memory.
```
```suggestion
- `ZMK_EV_EVENT_HANDLED`: The function received the event memory and has handled it. Stop propagating it to the next listeners and free the event payload memory.
```

henrebotha commented on docs/docs/development/new-behavior.md:
````diff
@@ -0,0 +1,381 @@
+
+The properties of the behavior are listed in the behavior's devicetree binding, which comes in the form of a `.yaml` file. Devicetree bindings are stored in the directory `app/dts/bindings/behaviors/` and are labelled in lowercase, beginning with the prefix `zmk,behavior-`, and ending with the behavior's name, using dashes to separate multiple words. For example, the directory for the hold-tap's  ...
+
+```
````
Speaking as a total outsider: Is it worth including syntax directives in the code snippets? At a glance the docs site doesn't have these, but Docusaurus does support it (as does GitHub).

````
```yaml
# SPDX-License-Identifier: MIT
description: Hold or Tap behavior
```
````

yields

```yaml
# SPDX-License-Identifier: MIT
description: Hold or Tap behavior
```

kurtis-lew replied:
The syntax-based colour-coding is definitely a nice indication for developers using this guide to confirm which parts should be edited versus those that should be left as-is. I'd be happy to implement the syntax directives in the upcoming commits. 👍

dxmh replied:
Definitely. I don’t see devicetree in [the list of supported languages for syntax highlighting](https://docusaurus.io/docs/markdown-features/code-blocks#supported-languages), but it would be nice to have highlighting wherever else possible. (We could maybe look to enable devicetree support for our docs in a separate PR?) 

Here’s the Docusaurus docs: [Code blocks](https://docusaurus.io/docs/markdown-features/code-blocks).

Also nice is [code titles](https://docusaurus.io/docs/markdown-features/code-blocks#code-title) and [line highlighting](https://docusaurus.io/docs/markdown-features/code-blocks#line-highlighting).

kurtis-lew commented:
@caksoylar @henrebotha thank you both as well for the extra review! A similar workflow has been implemented as those suggested by @dxmh: I'm committing the small changes now and will make the major changes at my earliest convenience.

caksoylar reviewed (approved):
Thanks for addressing all my comments!
